service/s3/s3crypto/cipher_util.go

```diff
@@ -0,0 +1,92 @@
+package s3crypto
+
+import (
+	"encoding/base64"
+	"strconv"
+
+	"github.com/aws/aws-sdk-go/aws/awserr"
+)
+
+func contentCipherFromEnvelope(env *Envelope, cfg Config) (ContentCipher, error) {
+	wrap, err := wrapFromEnvelope(env, cfg)
+	if err != nil {
+		return nil, err
+	}
+
+	return cekFromEnvelope(env, wrap)
+}
+
+func wrapFromEnvelope(env *Envelope, cfg Config) (CipherDataHandler, error) {
+	switch env.WrapAlg {
+	case "kms":
+		return NewKMSDecryptHandler(cfg.KMSSession, env.MatDesc)
+	}
+	return nil, awserr.New(
+		"InvalidWrapAlgorithmError",
+		"wrap algorithm isn't supported, "+env.WrapAlg,
+		nil,
+	)
+}
+
+// AESGCMNoPadding is the constant value that is used to specify
+// the CEK algorithm consiting of AES GCM with no padding.
+const AESGCMNoPadding = "AES/GCM/NoPadding"
+
+func cekFromEnvelope(env *Envelope, cdh CipherDataHandler) (ContentCipher, error) {
+	key, err := base64.StdEncoding.DecodeString(env.CipherKey)
+	if err != nil {
+		return nil, err
+	}
+
+	iv, err := base64.StdEncoding.DecodeString(env.IV)
+	if err != nil {
+		return nil, err
+	}
+	key, err = cdh.DecryptKey(key)
+	if err != nil {
+		return nil, err
+	}
+
+	switch env.CEKAlg {
+	case AESGCMNoPadding:
+		cd := CipherData{
+			Key: key,
+			IV:  iv,
+		}
+		return newAESGCMContentCipher(cd, cdh)
+	}
+	return nil, awserr.New(
+		"InvalidCEKAlgorithmError",
+		"cek algorithm isn't supported, "+env.CEKAlg,
+		nil,
+	)
+}
+
+func encodeMeta(reader HashReader, handler CipherDataHandler, cd *CipherData) (Envelope, error) {
```

Comment:
`CipherData` can stay a value an not a pointer since it does't need to be mutated.